### PR TITLE
Fixes #1038 If record type mismatch.

### LINKED ISSUE #1038: Crash from type mismatch
If() takes an intersection. 
Table() takes a union. 

Consider:
`If(false, { x : 1, y:1}, {x:1, z:2}).z`  is an error because If() takes the intersection. 

But 
`If(false, { x : 1, y:1}, {x:1, z:2})`

Still returns a runtime type showing {x:1, z:2} 
but compile time type  is {x}, so there's a mismatch between compile time and runtime type. 

PA Json() function correctly strips this out; but C# interpreter does not. 

Here are 2 more expressions that exploit this: 
```
First(Table(
  If(false, { x : 1, y:1}, {x:1, z:2}), // {x}, because intersection 
  {x:9,z:9})  // {x,z}
).z // Should be Blank()! actually showing 2
```
Z in first row should get stripped out .

If z in first row and z in second row have different types, leads to a crash in C# interpreter. 

```
Table(
If(false, { x : 1, y:1}, {x:1, z:"str"}), // {x}
{x:9,z:9}) // Blank(),  actually crashes 

```

### PR DESCRIPTION
**Pull request by jas-valgotar**
-Fixes #1038 
-This may wrap the result of the If statement in CompileTimeTypeWrapperRecordValue.

### CODE REVIEW

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/First.txt [hunk 1]
```diff
@@ -89,3 +89,8 @@ Table({Value:5})
 >> FirstN(If(false, Table({a:1}), Table({b:2})), 1)
 Table({})
 
+>> First(Table(If(false, {x:1, y:1}, {x:1, z:2}), {x:9,z:9})).z
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/First.txt:92]: Can you add a comment to help explain why. 
Actually, it'd even be fine to include this as a test case right before to help explain it. 

If(false, {x:1, y:1}, {x:1, z:2})

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/With.txt [hunk 2]
```diff
@@ -14,4 +14,4 @@ Blank()
 Error(Table({Kind:ErrorKind.Div0},{Kind:ErrorKind.Div0}))
 
 >> With({y:1}, If(false, {x:1}, {z:2}))
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/With.txt:16]: >>> [](http://example.com/codeflow?start=0&length=2)

nice - also worth commenting here.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs [hunk 3]
```diff
@@ -15,6 +15,7 @@
 using Microsoft.PowerFx.Core.Utils;
 using Microsoft.PowerFx.Interpreter;
 using Microsoft.PowerFx.Types;
+using static Microsoft.PowerFx.Syntax.PrettyPrintVisitor;
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs:18]: >PrettyPrintVisitor [](http://example.com/codeflow?start=38&length=18)

Do we need this?
PrettyPrinting seems suspicious  here

File: src/libraries/Microsoft.PowerFx.Core/Public/Values/CompileTimeTypeWrapperRecordValue.cs [hunk 4]
```diff
@@ -24,5 +27,44 @@ private CompileTimeTypeWrapperRecordValue(RecordType type, RecordValue inner)
             : base(IRContext.NotInSource(type), inner.Fields)
         {
         }
+
+        protected override bool TryGetField(FormulaType fieldType, string fieldName, out FormulaValue result)
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Public/Values/CompileTimeTypeWrapperRecordValue.cs:31]: >TryGetField [](http://example.com/codeflow?start=32&length=11)

Are these extra virtuals only required in the If() case where we *reduce* the type? (eg, have an intersection where the resulting type is smaller, and hence TryGetField needs to check if it should "block" a field?)

Ie, in Table() where we do a union, would we ever need these? My concern is the potential cost of the extra lookup. 
Either way - if it is a problem, I guess we'll find it in a profiler.

File: src/libraries/Microsoft.PowerFx.Core/Public/Values/CompileTimeTypeWrapperRecordValue.cs [hunk 5]
```diff
@@ -24,5 +27,44 @@ private CompileTimeTypeWrapperRecordValue(RecordType type, RecordValue inner)
...
+                // Only return field which were specified via the expectedType (IE RecordType),
+                // because inner record value may have more fields than the expected type.
+                return _fields.TryGetValue(fieldName, out result);
+            }
+
+            result = default;
+            return false;
+        }
+
+        public override async Task<DValue<RecordValue>> UpdateFieldsAsync(RecordValue changeRecord, CancellationToken cancellationToken)
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Public/Values/CompileTimeTypeWrapperRecordValue.cs:44]: >UpdateFieldsAsync [](http://example.com/codeflow?start=56&length=17)

Can this be merged with base class more?

File: src/libraries/Microsoft.PowerFx.Core/Public/Values/CompileTimeTypeWrapperRecordValue.cs (hunk 4 above)
Comment (jas-valgotar) [reply src/libraries/Microsoft.PowerFx.Core/Public/Values/CompileTimeTypeWrapperRecordValue.cs:31]: No this is only used for intersection.

File: src/libraries/Microsoft.PowerFx.Core/Public/Values/CompileTimeTypeWrapperRecordValue.cs (hunk 5 above)
Comment (jas-valgotar) [reply src/libraries/Microsoft.PowerFx.Core/Public/Values/CompileTimeTypeWrapperRecordValue.cs:44]: done.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/First.txt (hunk 1 above)
Comment (jas-valgotar) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/First.txt:93]: added

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/With.txt (hunk 2 above)
Comment (jas-valgotar) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/With.txt:17]: added

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by LucGenetier** [comment]
✅ No public API change.

**Review by MikeStall — APPROVED** [review]
:shipit:

**Comment by LucGenetier** [comment]
✅ No public API change.

**Pull request merged**